### PR TITLE
Taught pass manager to dump under debug log CSV-formatted events for compiler performance measurement.

### PR DESCRIPTION
**Pull request by mhong**
The flags to turn this on are `-Xllvm -sil-print-pass-time -Xllvm -debug-only=sil-passmanager`.

Example usage:
```
time ../build/$rdir/swift-linux-x86_64/bin/swiftc -O -Xllvm -sil-print-pass-time -Xllvm -debug-only=sil-passmanager some_code.swift 2>&1 | grep 'S4TF' >/tmp/log.csv
```

To import and analyze the CSV file in sqlite:

Run `sqlite3 profile.db`. In the `sqlite>` prompt, do the following:

```sql
drop table t;

CREATE TABLE t(
  "S4TF" TEXT,
  "Duration" INTEGER,
  "StartTime" TEXT,
  "PassName" TEXT,
  "PassType" TEXT,
  "FuncName" TEXT
);

.separator ","
.import /tmp/log.csv t

select count(*) from t;
```

Some example SQL queries for analysis:
```sql
-- total # of nanoseconds captured in the CSV events
select sum(Duration) from t;

-- top K funcs by cost
select sum(Duration), count(*),FuncName from t where passtype = 'F' group by 3 order by 1 desc;

-- top k passes for a particular function
select sum(Duration), count(*), PassName from t where FuncName = 'main' group by 3 order by 1 desc;
```

### CODE REVIEW

**Comment by mhong** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
Cool.  Can you propose this for upstream though?

**Comment by mhong** [comment]
SG. I'd like to land this patch in our branch first, and get some usage experiences. I might also add other stats logging features to our branch (e.g. to log the function code size before/after inlining), so that when we propose this upstream, we'll have a concrete set of "compiler performance analysis use cases" to back it up. Does this sound reasonable?

File: lib/SILOptimizer/PassManager/PassManager.cpp [hunk 1]
```diff
@@ -350,6 +350,17 @@ void SILPassManager::dumpPassInfo(const char *Title, unsigned TransIdx,
   llvm::dbgs() << '\n';
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+static void logS4TFPassEvent(llvm::sys::TimePoint<> StartTime, long long Delta,
+                             StringRef passName, bool isFunction,
```
Comment (bgogul) [inline lib/SILOptimizer/PassManager/PassManager.cpp]: nit: isFunction -> isFunctionPass?

File: lib/SILOptimizer/PassManager/PassManager.cpp [hunk 2]
```diff
@@ -415,6 +426,13 @@ void SILPassManager::runPassOnFunction(unsigned TransIdx, SILFunction *F) {
   if (SILPrintPassTime) {
     llvm::dbgs() << Delta << " (" << SFT->getID() << "," << F->getName()
                  << ")\n";
+
+    // SWIFT_ENABLE_TENSORFLOW
+    // Write CSV-formatted events, so that we can do aggregate analysis. Format:
+    // [S4TF] Delta,StartTime,PassName,PassType,FuncName
+    // Here PassType is F since it's a function pass.
+    logS4TFPassEvent(StartTime, Delta, SFT->getID(), /*isFunction*/ true,
```
Comment (bgogul) [inline lib/SILOptimizer/PassManager/PassManager.cpp]: nit: Swap StartTime & Delta to match the output?

**Review by bgogul — APPROVED** [review]
Just a couple of nits. 

Alternatively, can we get the csv by post-processing the sil-print-pass output instead of adding it here?

File: lib/SILOptimizer/PassManager/PassManager.cpp (hunk 1 above)
Comment (mhong) [reply lib/SILOptimizer/PassManager/PassManager.cpp]: Done

File: lib/SILOptimizer/PassManager/PassManager.cpp (hunk 2 above)
Comment (mhong) [reply lib/SILOptimizer/PassManager/PassManager.cpp]: Filed https://bugs.swift.org/browse/SR-9258 to track the upstreaming task.

**Comment by mhong** [comment]
Filed https://bugs.swift.org/browse/SR-9258 to track the upstreaming task.

**Comment by mhong** [comment]
@bgogul, I don't believe the existing debug log dumps time cost for each (function pass, SIL function) pair. It only dumps the aggregated time cost for a function pass.

**Comment by rxwei** [comment]
@mhong  This broke builds on macOS. 

```console
ptimizer/PassManager/PassManager.cpp
/usr/local/src/swift-build/swift/lib/SILOptimizer/PassManager/PassManager.cpp:357:50: error: no viable conversion from 'time_point<[...], duration<[...], ratio<[...], 1000000000>>>' to 'const time_point<[...], duration<[...], ratio<[...], 1000000>>>'
  auto tt = std::chrono::system_clock::to_time_t(StartTime);
                                                 ^~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/chrono:834:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'llvm::sys::TimePoint<>' (aka 'std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1, 1000000000> > >') to 'const std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1, 1000000> > > &' for 1st argument
class _LIBCPP_TEMPLATE_VIS time_point
                           ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/chrono:834:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'llvm::sys::TimePoint<>' (aka 'std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1, 1000000000> > >') to 'std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1, 1000000> > > &&' for 1st argument
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/chrono:853:5: note: candidate template ignored: requirement 'is_convertible<duration<long long, ratio<1, 1000000000> >, duration>::value' was not satisfied [with _Duration2 = std::__1::chrono::duration<long long, std::__1::ratio<1, 1000000000> >]
    time_point(const time_point<clock, _Duration2>& t,
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/chrono:1060:53: note: passing argument to parameter '__t' here
    static time_t     to_time_t  (const time_point& __t) _NOEXCEPT;
                                                    ^
1 error generated.
```

**Comment by mhong** [comment]
Thanks @rxwei! macOS CI was temporarily broken, so I did not block this PR on that. Sorry about the breakage!

**Pull request merged**